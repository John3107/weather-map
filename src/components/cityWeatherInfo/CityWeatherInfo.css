.cityInfo {
    display: flex;
    flex-direction: column;
    margin-top: 100px;
    width: 50%;
}

.date {
    display: inherit;
    justify-content: center;
    width: 100%;
    padding-bottom: 10px;
    border-bottom: #c7c3c3 solid 1px;
    border-bottom-left-radius: 0;
    border-bottom-right-radius: 0;
    font-family: sans-serif;
    font-size: 20px;
    margin-bottom: 20px;
}

.paperMain {
    display: inherit;
    flex-direction: column;
    padding: 10px;
}


.head {
    display: flex;
    flex-direction: column;
    align-items: center;
    width: 100%;
    margin-bottom: 20px;
}

.currentWeather {
    font-size: 12px;
    font-family: Solis Arial sans-serif;
    text-transform: uppercase;
    font-weight: 600;
}

.head img {
    width: 80px;
    height: 80px;
}

.body {
    width: 100%;
    display: flex;
    justify-content: space-between;
}

.side {
    display: inherit;
    flex-direction: column;
    width: 45%;
}

.param {
    display: inherit;
    justify-content: space-between;
    border-bottom: #c7c3c3 solid 1px;
    padding-bottom: 5px;
    margin-bottom: 5px;
    border-bottom-left-radius: 0;
    border-bottom-right-radius: 0;
}

@media (max-width: 600px) {
    .body {
        flex-direction: column;
        align-items: center;
    }

    .side {
        width: 100%;
    }
}

