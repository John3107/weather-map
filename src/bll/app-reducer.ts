import {CityType, CityWeatherType} from "../types/types";
import {AppThunk} from "./store";
import {weatherAPI} from "../api/api";

const initialState: InitialStateType = {
    cities: [],
    citiesContext: []
}

export const appReducer = (state: InitialStateType = initialState, action: ActionsType): InitialStateType => {
    switch (action.type) {
        case 'CITIES':
            return {...state, cities: action.citiesList}
        case 'CITIES-CONTEXT':
            return {...state, citiesContext: action.citiesContextArray}
        default:
            return state
    }
}

export const initialCitiesDataTC = (): AppThunk => (dispatch) => {
    const getCities = localStorage.getItem('cities-list')
    if (getCities) {
        dispatch(setCitiesAC(JSON.parse(getCities)))
    }
}

export const addCityTC = (newCity: CityType): AppThunk => (dispatch) => {
    dispatch(setCitiesContextAC([]))
    const citiesList = localStorage.getItem('cities-list')
    const citiesListParsed = citiesList && JSON.parse(citiesList)
    weatherAPI.getCityWeather(newCity.lat, newCity.lon)
        .then((res) => {
            const cityWeather = {
                id: String(res.data.lon) + String(res.data.lat),
                city: newCity.name,
                country: newCity.country,
                temp: Math.round(res.data.current.temp / 10.59),
                description: res.data.current.weather[0].description,
                icon: res.data.current.weather[0].icon,
                lat: res.data.lat,
                lon: res.data.lon
            }
            if (citiesListParsed[0]) {
                const searchCity = citiesListParsed.find((el: CityWeatherType) => el.id === cityWeather.id)
                let addCity
                if (searchCity) {
                    addCity = citiesListParsed.map((el: CityWeatherType) => el.id === cityWeather.id
                        ? ({...cityWeather})
                        : el)
                } else {
                    addCity = [...citiesListParsed, cityWeather]
                }
                localStorage.setItem('cities-list', JSON.stringify(addCity))
                dispatch(setCitiesAC(addCity))
            } else {
                localStorage.setItem('cities-list', JSON.stringify([cityWeather]))
                dispatch(setCitiesAC([cityWeather]))
            }
        })
        .catch((err) => {
            console.log(err)
        })
}

export const citiesContextTC = (title: string): AppThunk => (dispatch) => {
    if (!title) return dispatch(setCitiesContextAC([]))
    weatherAPI.getCitiesWeather(title)
        .then((res) => {
            dispatch(setCitiesContextAC(res.data))
        })
        .catch((err) => {
            console.log(err)
        })
}

export const deleteCardTC = (cardId: string): AppThunk => (dispatch) => {
    const citiesList = localStorage.getItem('cities-list')
    const citiesListParsed = citiesList && JSON.parse(citiesList)
    const citiesListFiltered = citiesListParsed.filter((el: CityWeatherType) => el.id !== cardId)
    localStorage.setItem('cities-list', JSON.stringify(citiesListFiltered))
    dispatch(setCitiesAC(citiesListFiltered))
}

export const setCitiesAC = (citiesList: CityWeatherType[]) => ({type: 'CITIES', citiesList} as const)
export const setCitiesContextAC = (citiesContextArray: CityType[]) =>
    ({type: 'CITIES-CONTEXT', citiesContextArray} as const)

export type CitiesActionType = ReturnType<typeof setCitiesAC>
export type CitiesContextActionType = ReturnType<typeof setCitiesContextAC>

export type InitialStateType = {
    cities: CityWeatherType[],
    citiesContext: CityType[]
}

type ActionsType = CitiesActionType | CitiesContextActionType