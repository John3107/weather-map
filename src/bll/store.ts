import {AnyAction, applyMiddleware, combineReducers, legacy_createStore as createStore} from 'redux'
import thunkMiddleware, {ThunkAction, ThunkDispatch} from 'redux-thunk'
import {appReducer} from "./app-reducer";

export const rootReducer = combineReducers({
    app: appReducer
})
export const store = createStore(rootReducer, applyMiddleware(thunkMiddleware));

export type AppRootStateType = ReturnType<typeof rootReducer>

export type AppDispatch = ThunkDispatch<AppRootStateType, unknown, AnyAction>

export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppRootStateType, unknown, AnyAction>