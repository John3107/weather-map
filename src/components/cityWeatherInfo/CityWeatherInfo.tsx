import React, {useEffect} from 'react';
import './CityWeatherInfo.css';
import {useAppDispatch, useAppSelector} from "../../hooks/hooks";
import {Paper} from "@material-ui/core";
import {selectedCityInitTC} from "../../bll/app-reducer";
import {ArrowUpward} from '@material-ui/icons';

const CityWeatherInfo = () => {
    const data = useAppSelector()
    const dispatch = useAppDispatch()

    useEffect(() => {
        dispatch(selectedCityInitTC())
    }, [dispatch])

    return (
        <div className="cityInfo">
            <div className="date">{data.selectedCity.date.split('GMT')[0]}</div>
            <Paper className="paperMain">
                <div className="head">
                    <span className="currentWeather">current weather</span>
                    <div className="main">
                        <img src={`http://openweathermap.org/img/wn/${data.selectedCity.icon}@2x.png`} alt=""/>
                        <span style={{fontSize: '30px'}}>{data.selectedCity.temp}C°</span>
                    </div>
                    <span>{data.selectedCity.description}</span>
                    <div>{data.selectedCity.name}, {data.selectedCity.country}</div>
                </div>
                <div className="body">
                    <div className="side">
                        <div className="param">
                            <span>Wind</span>
                            <span>{data.selectedCity.windSpeed} m/s</span>
                        </div>
                        <div className="param">
                            <span>Wind degree</span>
                            <ArrowUpward style={{transform: `rotate(${data.selectedCity.windDeg}deg)`}}/>
                        </div>
                        <div className="param">
                            <span>Humidity</span>
                            <span>{data.selectedCity.humidity}%</span>
                        </div>
                    </div>
                    <div className="side">
                        <div className="param">
                            <span>Dew Point</span>
                            <span>{data.selectedCity.dewPoint}C°</span>
                        </div>
                        <div className="param">
                            <span>Pressure</span>
                            <span>{data.selectedCity.pressure} mb</span>
                        </div>
                        <div className="param">
                            <span>Cloud Cover</span>
                            <span>{data.selectedCity.clouds}%</span>
                        </div>
                    </div>
                </div>
            </Paper>
        </div>
    );
}

export default CityWeatherInfo;