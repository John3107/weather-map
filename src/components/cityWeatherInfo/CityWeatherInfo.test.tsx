import { Provider } from "react-redux";
// @ts-ignore
import renderer from "react-test-renderer";
import CityWeatherInfo from './CityWeatherInfo';
import {store} from "../../bll/store";

it('renders correctly snapshot', () => {
    const tree = renderer
        .create(<Provider store={store}><CityWeatherInfo/></Provider>)
        .toJSON();
    expect(tree).toMatchSnapshot();
});