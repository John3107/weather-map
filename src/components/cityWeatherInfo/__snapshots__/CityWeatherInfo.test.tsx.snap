// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly snapshot 1`] = `
<div
  className="cityInfo"
>
  <div
    className="date"
  />
  <div
    className="MuiPaper-root paperMain MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="head"
    >
      <span
        className="currentWeather"
      >
        current weather
      </span>
      <div
        className="main"
      >
        <img
          alt=""
          src="http://openweathermap.org/img/wn/@2x.png"
        />
        <span
          style={
            Object {
              "fontSize": "30px",
            }
          }
        >
          0
          C°
        </span>
      </div>
      <span />
      <div>
        , 
      </div>
    </div>
    <div
      className="body"
    >
      <div
        className="side"
      >
        <div
          className="param"
        >
          <span>
            Wind
          </span>
          <span>
            0
             m/s
          </span>
        </div>
        <div
          className="param"
        >
          <span>
            Wind degree
          </span>
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            style={
              Object {
                "transform": "rotate(0deg)",
              }
            }
            viewBox="0 0 24 24"
          >
            <path
              d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
            />
          </svg>
        </div>
        <div
          className="param"
        >
          <span>
            Humidity
          </span>
          <span>
            0
            %
          </span>
        </div>
      </div>
      <div
        className="side"
      >
        <div
          className="param"
        >
          <span>
            Dew Point
          </span>
          <span>
            0
            C°
          </span>
        </div>
        <div
          className="param"
        >
          <span>
            Pressure
          </span>
          <span>
            0
             mb
          </span>
        </div>
        <div
          className="param"
        >
          <span>
            Cloud Cover
          </span>
          <span>
            0
            %
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
